app/services/fact_importer.rb

```diff
@@ -190,11 +190,16 @@ def normalize(facts)
 
   def db_facts
     query = host.fact_values
+    # MySQL query optimizer does not appear to pick the correct index here: https://projects.theforeman.org/issues/25053
     if ActiveRecord::Base.connection.adapter_name.downcase.starts_with? 'mysql'
-      # MySQL query optimizer does not appear to pick the correct index here: https://projects.theforeman.org/issues/25053
       query = query.from("fact_values USE INDEX(index_fact_values_on_fact_name_id_and_host_id)")
     end
-    query.where(:fact_name => fact_names.values).reorder('')
+    # When using PostgreSQL use regular expression to prevent query with thousands of parameters
+    if ActiveRecord::Base.connection.adapter_name.downcase.starts_with? 'postgresql'
+      query.where("fact_names.name ~ ?", "^(#{fact_names.values.map {|x| Regexp.escape(x)}.join('|')})$").reorder('')
+    else
+      query.where(:fact_name => fact_names.values).reorder('')
```

Comment:
I wonder, since most facts are repeated every time anyways, maybe it is best to not filter on fact names via the database at all here, but rather do it at https://github.com/theforeman/foreman/pull/6438/files#diff-671e5385f8ec83d391c87d7c292604edL174 where we go over the db_facts anyways?

Reply:
Good idea, I considered that but I quickly realized this would break the "contract" with plugins: Ansible and Salt both use the db_facts method. Let me think about this.

Reply:
We should have a fact source. Would that perhaps be the better thing to filter on? AFAIK the contract is that you provide a full set of facts for a fact source. We delete everything that's missing in the new set and add/update everything else.

Reply:
Fact source?

Reply:
We store if facts come from Puppet, Ansible or whatever. Could we use that to filter on instead?